src/gui/titlebar.cpp

```diff
@@ -0,0 +1,878 @@
+/*
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ *  Copyright (C) 2023-2024  The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "titlebar.h"
+
+#include "checks.h"
+#include "control.h"
+#include "cpu.h"
+#include "dosbox.h"
+#include "mapper.h"
+#include "sdlmain.h"
+#include "setup.h"
+#include "string_utils.h"
+#include "support.h"
+#include "video.h"
+
+#include <SDL.h>
+#include <vector>
+
+CHECK_NARROWING();
+
+// ***************************************************************************
+// Datat types and storage
+// ***************************************************************************
+
+static struct TitlebarConfig {
+	enum class ProgramDisplay { NotSet, Default, None, Name, Path, Segment, Custom };
+	enum class VersionDisplay { NotSet, Default, None, Simple, Detailed };
+	enum class MouseHintDisplay { NotSet, Default, None, Short, Full };
```

Comment:
Do we really need to differentiate between NotSet and None? If yes, a better solution would be not to encode the unset/uninitialised status in the enum values but use `std::optional`.

Reply:
There are two reasons:
- to warn about setting provided more than once
- in case of syntax error, to create a fixed string in `sync_config` which is similar to what the user originally provided (whether there was `mouse`, `mouse=full`, or nothing at all in the original string - I would like to preserve what the user submitted, despite every variant producing same results)

The second case can be handled in an alternative way, by storing the actual substring the user provided. Originally, I was considering this approach too - if you think it's better, I can change the implementation.
`std::optional` is not really suitable, as it would have turned a simple `switch (value)` into something more complex.

Reply:
Update: I'm going to store the original string. Having the animation set to enabled / set to disabled / not set at all / omitted makes some if's too complicated and the overall code would be too mistake-prone.

Reply:
Yeah I think storing the original string is the better solution @FeralChild64.

I did consider that using optional might overcomplicate the code. Now the switch block are exhaustive, so yeah, ifs would cause unnecesary complexity. As these enums are internal only, this is fine.

Reply:
Original settings are now stored in the strings.